### PR TITLE
Fix Ill-defined Definitions of Class/Interface

### PR DESCRIPTION
**Pull request by ZheyangSong**
+ The `Label` was ill-defined and didn't reflect the latest implementation
+ NOTE: this will cause breaking changes for any project that consumes the current `Label` interface.
  + The `Label` is used as a concrete class so that *<Label />* can be used in JSX
  + The `LabelProps` is redefined as what current `Label` interface.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not provide its own types, and you can not add them.
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
  - https://github.com/recharts/recharts/blob/03e554401a1b97e116746710247df0dd544b10dc/src/component/Label.js#L26
  - https://github.com/recharts/recharts/blob/03e554401a1b97e116746710247df0dd544b10dc/src/component/Label.js#L400-L401
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by mthmulders** [comment]
Thanks for your contribution, @ZheyangSong! Two questions from my side:

1. Your PR states that `Label` doesn't reflect the latest implementation. Can you please indicate when the implementation changed in this regard? Would it make sense to update the version number in the header?
1. The `parseViewBox` and `renderCallByParent` static methods that you have added to the `Label` class aren't listed in the [API docs](http://recharts.org/#/en-US/api/Label). As such, I assume they are not part of the public API. I'm not sure it makes sense to add them to the type definitions. @rapmue @royxue, how do you feel about that?

**Comment by ZheyangSong** [comment]
@mthmulders 
To the first question, a couple of things (such as `value` property was required in definition, but is optional in current document, `Label` should be an instantiable class rather than an interface and etc.) were ill-defined either initially or due to a later change. I need to check the repo's history to track down when the change happened. The last time I touched this one was a couple of months ago. I remember that this `Label` component was under development and consumed as a wrapper internally. I guess that's why it was defined as an interface. But the latest library exposes it already. As of bumping the version number, I looked up the releases in repo already. Problem is that `Recharts` team is releasing beta versions and DefinitelyTyped doesn't support beta versioning. That's why the version of this definition file hasn't been changed for a long time, though the various updates happened.

To the second question, the changed definition doesn't exactly follow the API docs on Recharts' site. It's based on the source code directly. I agree with you on sticking to the API docs. I originally thought that these two methods might be needed for advanced usage (such as customizing the positioning of the Label in some custom component)...

**Comment by mthmulders** [comment]
1. Just saw that in the header of the current type definitions, it already says "1.0". This is strange, as it isn't currently released. But well, since the type defs are at 1.0 already, I think the version numbering shouldn't be an issue. Since the type defs are "pre-1.0", the refactored Label should be a good fit in it, too.
1. I would prefer to stick with the documented API on the Recharts website.  So I'd like to have the undocumented methods removed.

**Comment by ZheyangSong** [comment]
@mthmulders Okay. Let me remove the two methods.

File: types/recharts/index.d.ts [hunk 1]
```diff
@@ -821,22 +821,35 @@ export interface TreemapProps extends EventAttributes, Animatable {
...
+    top: number;
+    left: number;
+    width: number;
+    height: number;
+    clockWise: boolean;
+}
+
+export class Label extends React.Component<LabelProps> { }
+
+export interface LabelProps {
```
Comment (mthmulders) [inline types/recharts/index.d.ts]: Could you please also add an example of how to properly use `Label` in recharts-tests.tsx? That way people will understand how they can use it themselves. Thanks!

**Review by mthmulders — CHANGES_REQUESTED** [review]

File: types/recharts/index.d.ts (hunk 1 above)
Comment (ZheyangSong) [reply types/recharts/index.d.ts]: @mthmulders updated `recharts-test.tsx`. I think the official documentation is more reliable for developers.

Comment (mthmulders) [reply types/recharts/index.d.ts]: Thanks a lot, looks good to me!

File: types/recharts/index.d.ts [hunk 2]
```diff
@@ -821,22 +821,35 @@ export interface TreemapProps extends EventAttributes, Animatable {
 
 export class Treemap extends React.Component<TreemapProps> { }
 
-export interface Label {
-    className?: string;
+export interface parseViewBoxArgs {
```
Comment (mthmulders) [inline types/recharts/index.d.ts]: Am I correct that this interface is no longer needed?

**Review by mthmulders — CHANGES_REQUESTED** [review]

File: types/recharts/index.d.ts (hunk 2 above)
Comment (ZheyangSong) [reply types/recharts/index.d.ts]: @mthmulders good catch. Thanks.

**Review by mthmulders — APPROVED** [review]

**Pull request merged**